### PR TITLE
[Merged by Bors] - feat: improve `polyrith` by testing for membership in the radical

### PR DESCRIPTION
**Pull request by robertylewis**
In @hrmacbeth 's tutorial on `polyrith`, there were examples of problems that it could almost solve, but failed. The goal was not expressible as a linear combination of the hypotheses but a power of the goal was. 

```lean
example (x y z : ℚ) (h : x = y) (h2 : x * y = 0) : x + y*z = 0 :=
sorry
```

Mathematically, `x+y*z` is in the radical of the ideal generated by `x-y, x*y`. There's a "standard trick" for testing membership in the radical without a search for the proper exponent: see e.g. section 2.2 of [arxiv.org/pdf/1007.3615.pdf](https://arxiv.org/pdf/1007.3615.pdf) or 4.2 Prop 8 of Cox, Little, O'Shea.

This PR implements the trick in the Sage call made by `polyrith`. When the power returned is `n > 1`, we use `linear_combination (exp := n)` to check the certificate (#7789 ).

The `polyrith` test infrastructure still needs to be ported from mathlib3. All tests in the test file succeed when they are uncommented. A future PR will restore the old test suite.

### CODE REVIEW

File: Mathlib/Tactic/Polyrith.lean [hunk 1]
```diff
@@ -250,14 +250,20 @@ instance : FromJson Poly where
         mon := mon.mul' (.pow' (← fromJson? (← j.getArrVal? 0)) (← fromJson? (← j.getArrVal? 1)))
       pure mon
 
+/-- A schema for the data reported by the Sage calculation -/
+structure SageCoeffAndPower where
+  /-- The function call produces an array of polynomials
+  parallel to the input list of hypotheses. -/
+  coeffs : Array Poly
+  power  : ℕ
```
Comment (robertylewis) [inline Mathlib/Tactic/Polyrith.lean:258]:
```suggestion
  /-- Sage produces an exponent (default 1) in the case where the hypothesess
  sum to a power of the goal. -/
  power  : ℕ
```

File: Mathlib/Tactic/Polyrith.lean [hunk 2]
```diff
@@ -250,14 +250,20 @@ instance : FromJson Poly where
...
+
 /-- The result of a sage call in the success case. -/
 structure SageSuccess where
   /-- The script returns a string containing python script to be sent to the remote server,
   when the tracing option is set. -/
   trace : Option String := none
-  /-- The main result of the function call is an array of polynomials
-  parallel to the input list of hypotheses. -/
-  data : Option (Array Poly) := none
+  data : Option SageCoeffAndPower := none
```
Comment (robertylewis) [inline Mathlib/Tactic/Polyrith.lean:266]:
```suggestion
  /-- The main result of the function call is an array of polynomials
  parallel to the input list of hypotheses and an exponent for the goal. -/
  data : Option SageCoeffAndPower := none
```

**Comment by digama0** [comment]
Note: I have pushed an update to the lean toolchain because this PR was on a [buggy version](https://leanprover.zulipchat.com/#narrow/stream/113488-general/topic/Potential.20data.20loss.20from.20.60lake.20clean.60.20with.204.2E2.2E0-rc2.2F3/near/397875701) of the toolchain. **WARNING**: checking out old commits of this PR using v4.2.0-rc2 or v4.2.0-rc3 can cause `lake clean` to delete your mathlib folder! If you need to do so, make sure to delete `lakefile.olean` manually before running any `lake` commands.

**Comment by adomani** [comment]
The code is above what I actually feel that I understand, but is this testing membership in the radical or whether the some explicit power of the element lies in the ideal?  More explicitly, could it not be the case that a larger power than what is being used would work?

As I said, I do not really understand the code, so I hope that this question is not just noise!

**Comment by robertylewis** [comment]
It tests for membership in the radical. 

To start, the goal is `H₁ = 0, ..., Hₙ = 0 ⊢ P = 0`.  This follows if we can find `Qᵢ` and `r` such that `Pʳ = Q₁H₁ + ... + QₙHₙ`, i.e. `P` is in the radical of `<Hᵢ>`.

In Sage: introduce a new variable `z` and use a Grobner basis algorithm to compute `hᵢ` such that `1 = h₁H₁ + ... + hₙHₙ + h(1 - zP)`. 

Take `r` to be the maximum power of `z` that appears in the `hᵢ`s. Substitute `1/P` (formally) for `z` and multiply `1 = h₁H₁ + ... + hₙHₙ + h(1 - zP)` by `Pʳ` to get our desired equation: the coeffs are `Qᵢ = Pʳhᵢ[z ← 1/P]`.
 
So the Grobner basis computation produces `r` without a search. In principle, if there is an `r` that works, this should find it. Of course it may be computationally infeasible.

I'm not sure that this procedure makes any claim about finding a *minimal* `r`. This must depend on the details of the Grobner basis algorithm. In practice, on all the `polyrith` tests, it correctly finds `r = 1` (and finds a minimal `r > 1` on the new tests).

**Comment by adomani** [comment]
Robert, thank you for the explanation!  I had missed that there was a first Gröbner basis computation and then you used the degrees of the elements of that basis to get the bound on `r`: this clears my doubts!

In terms of improving feasibility, it may be good to have an option to ask Sage to check if the given element is a member of the ideal itself or whether it is contained in the radical: I expect these two computations to perform differently (where testing for membership in the ideal is likely often faster) and you may be satisfied with the answer to one of the two questions.

Anyway, this is just a suggestion: I am very happy that `polyrith` even exists!

**Comment by robertylewis** [comment]
I thought about the two versions, but decided against it for a few reasons:

* It's confusing for newcomers/people who don't want to think about how `polyrith` works, which is probably most people. They just want to know whether the goal is provable.
* `polyrith` is self-replacing, so minor slowdowns don't matter nearly as much as for tactics that will remain source code.
* It's not even clear to me that the radical problem is significantly slower on average. The computation has one more variable and similar degree. Grobner basis algorithms are fickle and adding an option to use a different term order would probably help a lot more than falling back to the original non-radical check.

**Comment by adomani** [comment]
Rob, thanks for the further comments!  I agree that specifying term orders normally has a *much* larger impact on feasibility of a Gröbner basis computation!

Thanks for taking the time to explain: it turns out that my comments were indeed noise, but I am grateful for the explanations, since they taught me something!

File: scripts/polyrith_sage.py [hunk 3]
```diff
@@ -19,19 +19,26 @@ def create_query(type: str, n_vars: int, eq_list, goal_type):
...
-    [{var_list}] = P.gens()
+    P = PolynomialRing({type_str(type)}, {var_list})
+    [{", ".join(var_list)}] = P.gens()
+    p = P({goal_type})
+    gens = {eq_list} + [1 - p*aux]
+    I = P.ideal(gens)
+    coeffs = P(1).lift(I)
+    power = max(cf.degree(aux) for cf in coeffs)
+    coeffs = [P(cf.subs(aux = 1/p)*p^power) for cf in coeffs[:int(-1)]]
+    print(str(power)+';'+serialize_polynomials(coeffs))
```
Comment (alexjbest) [inline scripts/polyrith_sage.py:33]: I don't really understand why this is done with this trick of checking if 1 is in the ideal `<hyps, 1/goal>`, which, I would expect singular/sage has a way to ask for a lift to the radical ideal directly which I would hope would be easier to maintain in the long run (especially if we have aspirations of making this work for non-char 0 rings properly). Did you consider this approach?
I'm not completely against merging this, but the whole thing is starting to look a bit scary ;)

File: scripts/polyrith_sage.py [hunk 4]
```diff
@@ -19,19 +19,26 @@ def create_query(type: str, n_vars: int, eq_list, goal_type):
...
     # `TypeError: no conversion of this ring to a Singular ring defined`
     P = PolynomialRing({type_str(type)}, 'var', 1)
-p = P({goal_type})
-I = P.ideal({eq_list})
-coeffs = p.lift(I)
-print(serialize_polynomials(coeffs))
+    p = P({goal_type})
+    I = P.ideal({eq_list})
+    coeffs = p.lift(I)
+    print('1;'+serialize_polynomials(coeffs))
```
Comment (eric-wieser) [inline scripts/polyrith_sage.py:41]: Can you add a comment explaining why this needs to be in the "workaround" branch?

File: scripts/polyrith_sage.py (hunk 3 above)
Comment (alexjbest) [reply scripts/polyrith_sage.py:33]: After experimenting a bit I believe that using the builtin radical support, while probably a better algorithm underneath, is not very convenient to use with the way things are set up here, so I'm happy with this I guess

Comment (robertylewis) [reply scripts/polyrith_sage.py:33]: Yeah, I remember looking for this when I first implemented this (granted, a year and a half ago) and not turning up anything useful. Sage/Singular will check for membership in the radical just fine but extracting the power and coefficient is a pain. My impression is that what I do in this PR is a "standard" approach and that a native Sage implementation would do roughly the same, although I'm somewhat out of my element here.

File: scripts/polyrith_sage.py (hunk 4 above)
Comment (robertylewis) [reply scripts/polyrith_sage.py:44]: Done!

File: scripts/polyrith_sage.py (hunk 3 above)
Comment (alexjbest) [reply scripts/polyrith_sage.py:33]: When I read the sage docstring for radical earlier it said:
```
             From the Singular manual: A combination of the algorithms
            of Krick/Logar and Kemper is used. Works also in positive
            characteristic (Kempers algorithm).
```
but I have no idea how accurate that is!

**Comment by alexjbest** [comment]
maintainer merge

File: scripts/polyrith_sage.py (hunk 3 above)
Comment (robertylewis) [reply scripts/polyrith_sage.py:33]: Right -- but there's space between the membership check $p \in \sqrt{\langle H_1, \ldots, H_n \rangle}$, and explicitly producing $k$ and $Q_i$ such that $p^k = \sum Q_iH_i$. Sage's `radical` makes the former easy, but I don't see how to get from there to the latter. Maybe there's a way, I'd love to hear it.

(It sounds like you came to a similar conclusion, just leaving this comment for reference later!)

**Comment by robertylewis** [comment]
bors merge

**Pull request closed without merging**